src/sentry/event_manager.py

```diff
@@ -608,18 +608,42 @@ def save(self, project, raw=False):
         # store a reference to the group id to guarantee validation of isolation
         event.data.bind_ref(event)
 
-        try:
-            with transaction.atomic(using=router.db_for_write(EventMapping)):
-                EventMapping.objects.create(project=project, group=group, event_id=event_id)
-        except IntegrityError:
+        # When an event was sampled, the canonical source of truth
+        # is the EventMapping table since we aren't going to be writing out an actual
+        # Event row. Otherwise, if the Event isn't being sampled, we can safely
+        # rely on the Event table itself as the source of truth and ignore
+        # EventMapping since it's redundant information.
+        if is_sample:
+            try:
+                with transaction.atomic(using=router.db_for_write(EventMapping)):
+                    EventMapping.objects.create(project=project, group=group, event_id=event_id)
+            except IntegrityError:
+                self.logger.info(
+                    'duplicate.found',
+                    exc_info=True,
+                    extra={
+                        'event_uuid': event_id,
+                        'project_id': project.id,
+                        'group_id': group.id,
+                        'model': EventMapping.__name__,
+                    }
+                )
+                return event
+
+        # We now always need to check the Event table for dupes
+        # since EventMapping isn't exactly the canonical source of truth.
+        if Event.objects.filter(
+            project_id=project.id,
+            event_id=event_id,
+        ).exists():
             self.logger.info(
                 'duplicate.found',
                 exc_info=True,
                 extra={
                     'event_uuid': event_id,
                     'project_id': project.id,
                     'group_id': group.id,
-                    'model': EventMapping.__name__,
+                    'model': Event.__name__,
                 }
             )
             return event
```

Comment:
I didn't realize `EventManager.save` could return an event that … wasn't saved. 😶

Reply:
Hi @tkaemming.

Reply:
Hi